query-service-impl/src/main/java/org/hypertrace/core/query/service/AdditionalHandlerFiltersConfig.java

```diff
@@ -0,0 +1,128 @@
+              .collect(
+                  Collectors.toMap(
+                      tenantFilters -> tenantFilters.tenantId,
+                      tenantFilters -> tenantFilters.filters));
+    } else {
+      this.tenantToAdditionalFiltersMap = Collections.emptyMap();
+    }
+  }
+
+  public List<Filter> getAdditionalFiltersForTenant(String tenantId) {
+    return this.tenantToAdditionalFiltersMap.getOrDefault(tenantId, Collections.emptyList());
+  }
+
+  @lombok.Value
+  @NonFinal
+  private static class TenantFilters {
+    private static final String TENANT_ID_CONFIG_KEY = "tenantId";
+    private static final String TIME_RANGE_AND_FILTERS_CONFIG_KEY = "timeRangeAndFilters";
+    String tenantId;
+    List<Filter> filters;
+
+    private TenantFilters(Config config, Optional<String> startTimeAttributeName) {
+      this.tenantId = config.getString(TENANT_ID_CONFIG_KEY);
+      this.filters =
+          config.getConfigList(TIME_RANGE_AND_FILTERS_CONFIG_KEY).stream()
+              .map(TimeRangeAndFilter::new)
+              .map(
+                  timeRangeAndFilter ->
+                      timeRangeAndFilter.buildTimeRangeAndFilters(startTimeAttributeName))
+              .filter(Optional::isPresent)
+              .map(Optional::get)
+              .collect(Collectors.toList());
+    }
+  }
+
+  @lombok.Value
+  @NonFinal
+  private static class TimeRangeAndFilter {
+    private static final String START_TIME_CONFIG_PATH = "startTimeMillis";
+    private static final String END_TIME_CONFIG_PATH = "endTimeMillis";
+    private static final String FILTER_CONFIG_PATH = "filter";
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+    Optional<Long> startTimeMillis;
+    Optional<Long> endTimeMillis;
+    Optional<Filter> filter;
+
+    private TimeRangeAndFilter(Config config) {
+      if (config.hasPath(START_TIME_CONFIG_PATH) && config.hasPath(END_TIME_CONFIG_PATH)) {
+        this.startTimeMillis = Optional.of(config.getLong(START_TIME_CONFIG_PATH));
+        this.endTimeMillis = Optional.of(config.getLong(END_TIME_CONFIG_PATH));
+      } else {
+        startTimeMillis = Optional.empty();
+        endTimeMillis = Optional.empty();
+      }
+      if (config.hasPath(FILTER_CONFIG_PATH)) {
+        this.filter = deserializeFilter(config.getString(FILTER_CONFIG_PATH));
+      } else {
+        filter = Optional.empty();
+      }
+    }
+
+    private Optional<Filter> buildTimeRangeAndFilters(Optional<String> timeRangeAttributeOptional) {
+      Filter.Builder filterBuilder = Filter.newBuilder();
+      filterBuilder.setOperator(Operator.OR);
+      if (timeRangeAttributeOptional.isPresent()) {
+        if (this.startTimeMillis.isPresent() && this.endTimeMillis.isPresent()) {
+          filterBuilder.addChildFilter(
+              QueryRequestUtil.createLongFilter(
+                  timeRangeAttributeOptional.get(), Operator.GT, this.endTimeMillis.get()));
```

Comment:
Q: Is this check correct? Should `endTime` be compared with the greater than operator (I don't remember the text you had in your sublime, so just re-checking)?

Reply:
Yes this is how it's supposed to be as it's an or filter 
This is what we discussed.

```
t3....   t1  t2 ....t4


((s >= t2 OR s <= t1) OR (API NOT IN (api1, api2)) AND (t3, t4))
```